cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+      if (parquet_types[equality_col_idx] == Type::INT96) {
+        // For INT96 timestamps, use cudf::string_view type and set timestamp_is_int96
+        return query_bloom_filter<cudf::string_view, true>(
+          equality_col_idx, dtype, literal, stream, mr);
+      } else {
+        return query_bloom_filter<T>(equality_col_idx, dtype, literal, stream, mr);
+      }
+    } else {
+      return query_bloom_filter<T>(equality_col_idx, dtype, literal, stream, mr);
+    }
+  }
+};
+
+/**
+ * @brief Collects lists of equality predicate literals in the AST expression, one list per input
+ * table column. This is used in row group filtering based on bloom filters.
+ */
+class equality_literals_collector : public ast::detail::expression_transformer {
+ public:
+  equality_literals_collector() = default;
+
+  equality_literals_collector(ast::expression const& expr, cudf::size_type num_input_columns)
+    : _num_input_columns{num_input_columns}
+  {
+    _equality_literals.resize(_num_input_columns);
+    expr.accept(*this);
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::literal const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::literal const& expr) override
+  {
+    _bloom_filter_expr = std::reference_wrapper<ast::expression const>(expr);
+    return expr;
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::column_reference const& expr) override
+  {
+    CUDF_EXPECTS(expr.get_table_source() == ast::table_reference::LEFT,
+                 "BloomfilterAST supports only left table");
+    CUDF_EXPECTS(expr.get_column_index() < _num_input_columns,
+                 "Column index cannot be more than number of columns in the table");
+    _bloom_filter_expr = std::reference_wrapper<ast::expression const>(expr);
+    return expr;
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_name_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(
+    ast::column_name_reference const& expr) override
+  {
+    CUDF_FAIL("Column name reference is not supported in BloomfilterAST");
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::operation const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::operation const& expr) override
+  {
+    using cudf::ast::ast_operator;
+    auto const operands = expr.get_operands();
+    auto const op       = expr.get_operator();
+
+    if (auto* v = dynamic_cast<ast::column_reference const*>(&operands[0].get())) {
+      // First operand should be column reference, second should be literal.
+      CUDF_EXPECTS(cudf::ast::detail::ast_operator_arity(op) == 2,
+                   "Only binary operations are supported on column reference");
+      CUDF_EXPECTS(dynamic_cast<ast::literal const*>(&operands[1].get()) != nullptr,
```

Comment:
Save the result of this `dynamic_cast` to a variable before checking if it's `nullptr`, and then reuse it later in the `emplace_back` call.